packages/panzer/disc-fe/src/evaluators/Panzer_ScatterDirichletResidual_Tpetra_impl.hpp

```diff
@@ -332,15 +417,111 @@ preEvaluate(typename TRAITS::PreEvalData d)
   std::vector<std::string> activeParameters =
     rcp_dynamic_cast<ParameterList_GlobalEvaluationData>(d.gedc->getDataObject("PARAMETER_NAMES"))->getActiveParameters();
 
-  // ETP 02/03/16:  This code needs to be updated to properly handle scatterIC_
-  TEUCHOS_ASSERT(!scatterIC_);
-  dfdp_vectors_.clear();
+  dfdpFieldsVoV_.initialize("ScatterResidual_Tpetra<Tangent>::dfdpFieldsVoV_",activeParameters.size());
+
   for(std::size_t i=0;i<activeParameters.size();i++) {
     RCP<typename LOC::VectorType> vec =
       rcp_dynamic_cast<LOC>(d.gedc->getDataObject(activeParameters[i]),true)->get_f();
-    Teuchos::ArrayRCP<double> vec_array = vec->get1dViewNonConst();
-    dfdp_vectors_.push_back(vec_array);
+    auto dfdp_view = vec->getLocalViewDevice(Tpetra::Access::ReadWrite);
+
+    dfdpFieldsVoV_.addView(dfdp_view,i);
+  }
+
+  dfdpFieldsVoV_.syncHostToDevice();
+
+}
+
+// **********************************************************************
+namespace panzer {
+namespace {
+
+template <typename ScalarT,typename LO,typename GO,typename NodeT>
+class ScatterDirichletResidual_Tangent_Functor {
+public:
+  typedef typename PHX::Device execution_space;
+  typedef PHX::MDField<const ScalarT,Cell,NODE> ScalarFieldType;
+  typedef PHX::MDField<const bool,Cell,NODE> BoolFieldType;
+
+  Kokkos::View<double**, Kokkos::LayoutLeft,PHX::Device> r_data;
+  Kokkos::View<double**, Kokkos::LayoutLeft,PHX::Device> dirichlet_counter;
+
+  Kokkos::View<Kokkos::View<double**,Kokkos::LayoutLeft,PHX::Device>*>  dfdp_fields; // tangent fields
+  double num_params;
+
+  PHX::View<const LO**> lids;    // local indices for unknowns
+  PHX::View<const int*> offsets; // how to get a particular field
+  PHX::View<const int*> basisIds;
+  ScalarFieldType field;
+  BoolFieldType applyBC;
+
+  bool checkApplyBC;
+
+  KOKKOS_INLINE_FUNCTION
+  void operator()(const unsigned int cell) const
+  {
+
+    // loop over the basis functions (currently they are nodes)
+    for(std::size_t basis=0; basis < offsets.extent(0); basis++) {
+       int offset = offsets(basis);
+       LO lid    = lids(cell,offset);
+       if (lid<0) continue; // not on this processor
+
+       int basisId = basisIds(basis);
+       if (checkApplyBC)
+         if(!applyBC(cell,basisId)) continue;
```

Comment:
@rppawlo Also, not sure if we like this branching within the parallel_for. I saw it elsewhere so may not be a concern.